linode/terraform-provider-linode
# new: Add `linode_placement_group_assignment` resource

## 📝 Description

This pull request adds a new `linode_placement_group_assignment` resource, which manages a single assignment between a Placement Group and a Linode.

Additionally, this PR adds a new `placement_group_externally_managed` attribute to the `linode_instance` resource which disables diffs on the `placement_group` field when enabled.

NOTE: Because this resource does not correspond to a single API resource, the ID attribute is stored as a base64-encoded JSON string containing the Linode ID and PG ID. 

## ✔️ How to Test

The following test steps assume you have pulled down this PR locally and pointed your local environment at an API instance where PGs are available:

```
export LINODE_URL=...
export LINODE_TOKEN=...
```

### Integration Testing

```
make PKG_NAME=linode/placementgroupassignment int-test
make PKG_NAME=linode/instance int-test
```

### Manual Testing

1. In a Terraform provider sandbox environment (e.g. dx-devenv), apply the following configuration:

```terraform
resource "linode_placement_group_assignment" "my-assignment" {
  placement_group_id = linode_placement_group.my-pg.id
  linode_id = linode_instance.my-inst.id
}

resource "linode_placement_group" "my-pg" {
  label = "my-pg"
  region = "us-east"
  affinity_type = "anti_affinity:local"
}

resource "linode_instance" "my-inst" {
  label = "my-inst"
  region = "us-east"
  type = "g6-nanode-1"
  placement_group_externally_managed = true
}
```

2. Ensure the apply finishes successfully.
3. Navigate to the newly created Placement Group in Cloud Manager and ensure the new instance has been assigned to it.
4. Remove the Placement Group assignment from the Terraform state:

```
terraform state rm linode_placement_group_assignment.my-assignment
```

5. Run the following command to re-import the previously created Placement Group:

```
make import ARGS="linode_placement_group_assignment.my-assignment MYPGID,MYINSTANCEID"
```

6. Re-apply the configuration and ensure no changes are proposed.
7. Comment out the `linode_placement_group_assignment` resource in the configuration and re-apply.
8. Ensure the apply finishes successfully.
9. Navigate to the Placement Group in Cloud Manager and ensure the Placement Group does not have any assigned Linodes.
